domains/transfer/TransferAgreementAutomation/Worker/Service/Engine/TransferCertificatesBasedOnConsumptionEngine.cs

```diff
@@ -0,0 +1,154 @@
+    private readonly ILogger<TransferCertificatesBasedOnConsumptionEngine> _logger;
+    private readonly IProjectOriginWalletClient _walletClient;
+
+    public TransferCertificatesBasedOnConsumptionEngine(IRequestStatusRepository requestStatusRepository, TransferEngineUtility transferUtility,
+        ILogger<TransferCertificatesBasedOnConsumptionEngine> logger, IProjectOriginWalletClient walletClient)
+    {
+        _requestStatusRepository = requestStatusRepository;
+        _transferUtility = transferUtility;
+        _logger = logger;
+        _walletClient = walletClient;
+    }
+
+    public bool IsSupported(TransferAgreement transferAgreement)
+    {
+        return transferAgreement.Type == TransferAgreementType.TransferCertificatesBasedOnConsumption;
+    }
+
+    public async Task TransferCertificates(TransferAgreement transferAgreement, CancellationToken cancellationToken = default)
+    {
+        if (!IsSupported(transferAgreement))
+        {
+            throw new ArgumentException(nameof(transferAgreement));
+        }
+
+        var senderId = transferAgreement.SenderId;
+        var receiverId = transferAgreement.ReceiverId;
+
+        if (receiverId is null)
+        {
+            _logger.LogInformation("Skipping consumption based transfer agreement {Id}, sender id is not specified", transferAgreement.Id);
+            return;
+        }
+
+        if (await _transferUtility.HasPendingTransactions(senderId, cancellationToken))
+        {
+            _logger.LogInformation("Skipping transfer agreement {Id}, sender {OrgId} has pending transactions", transferAgreement.Id, senderId);
+            return;
+        }
+
+        if (await _transferUtility.HasPendingTransactions(receiverId, cancellationToken))
+        {
+            _logger.LogInformation("Skipping transfer agreement {Id}, receiver {OrgId} has pending transactions", transferAgreement.Id, senderId);
+            return;
+        }
+
+        var receiverCertificates = await _transferUtility.GetCertificates(receiverId);
+        _logger.LogInformation("Found {Count} certificates for receiver", receiverCertificates.Count);
+
+        if (receiverCertificates.Any(c => c.CertificateType == CertificateType.Consumption))
+        {
+            var receiverUnmatchedConsumptionGroupedByPeriod = GetUnmatchedConsumptionGroupedByPeriod(receiverCertificates);
+
+            _logger.LogInformation("Found {Count} periods where receiver has unmatched consumption",
+                receiverUnmatchedConsumptionGroupedByPeriod.Count);
+
+            var senderCertificates = await _transferUtility.GetCertificates(senderId);
+            var senderProductionCertificatesGroupedByPeriod = GetProductionCertificatesGroupedByPeriod(senderCertificates);
+
+            foreach (var receiverGroup in receiverUnmatchedConsumptionGroupedByPeriod)
+            {
+                var senderCertificatesInPeriod = senderProductionCertificatesGroupedByPeriod
```

Comment:
This has some complexity. Would love if this could be made simpler somehow :/

Reply:
I hope is a bit easier to read now :)